message/messaging_service.cc

```diff
@@ -800,12 +800,13 @@ shared_ptr<messaging_service::rpc_protocol_client_wrapper> messaging_service::ge
         opts.isolation_cookie = _scheduling_info_for_connection_index[idx].isolation_cookie;
     }
 
-    auto baddr = socket_address(utils::fb_utilities::get_broadcast_address(), 0);
+    bool listen_to_bc = _cfg.listen_on_broadcast_address && _cfg.ip != utils::fb_utilities::get_broadcast_address();
```

Comment:
From listen_on_broadcast_address comment:

> When using multiple physical network interfaces, set this to true to listen on broadcast_address in addition to the listen_address, allowing nodes to communicate in both interfaces.  Ignore this property if the network configuration automatically routes between the public and private networks such as EC2.

So it looks like when that is true, we need to listen on both. And when false, we only need to listen on listen_address.

(The `true` setting is useful when you want both internal clients and external clients)

Reply:
Yes. We do that. And in this case, the actual behaviour when doing dc/rack encryption will depend on which of the two interfaces broadcast is. Here (in this connect bind) we opt to locally bind to broadcast in this case, since it is proven that this is bindable (we listen there), and we sort of hope that this is what gossip will indicate. But one could argue that in the case of not encrypting, for example rack, the "listen_address" would be "more private" and possibly where the non-tls connections come, so maybe this is bad instead. I seriously can't tell which.

Reply:
Ahh, I missed that it's the client side. But why bind at all? Let the kernel figure it out.

Reply:
Again, because of #8051. We need the connecting side of rpc to have an address that is in the "who is in the dc/rack" info propagated by gossip. This is typically actually broadcast address (where call go to). Now it turns out this does not work well in AWS env, so we can hope using listen address will have much the same effect in this and other cases, and NAT fixing connections looking like they are where we hope they are from.

Reply:
Ah, so we bind to our listen_address but send to the other node's broadcast_address, hoping NAT will translate the other node's broadcast_address to its listen_address.

In numbers to make it clearer (for me at least):

 N1=10.0.0.1/211.0.0.1 (say we own 211.0.0.0/8).
 N2=10.0.0.2/211.0.0.2

N2 listens on 10.0.0.2 and broadcasts 211.0.0.2
N1 binds to 10.0.0.1 and connects to 211.0.0.2
NAT (we hope) changes the connection to 10.0.0.1->10.0.0.2

Is this right?

If so, it must be tested in a realistic environment.

Reply:
Correct, and yes. Somewhat complicated by my general AMI ignorance...